crates/rustc_codegen_spirv/src/linker/destructure_composites.rs

```diff
@@ -0,0 +1,96 @@
+//! Simplify `OpCompositeExtract` pointing to `OpCompositeConstruct`s / `OpCompositeInsert`s.
+//! Such constructions arise after inlining, when using multi-argument closures
+//! (and other `Fn*` trait implementations). These composites can frequently be invalid,
+//! containing pointers, `OpFunctionArgument`s, etc. After simplification, components
+//! will become valid targets for `OpLoad`/`OpStore`.
+use super::apply_rewrite_rules;
+use rspirv::dr::{Instruction, Module, Operand};
+use rspirv::spirv::Op;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+
+pub fn destructure_composites(module: &mut Module) {
+    for function in module.functions.iter_mut() {
+        let mut rewrite_rules = FxHashMap::default();
+        let reference: FxHashMap<_, _> = function
+            .all_inst_iter()
+            .filter_map(|inst| match inst.class.opcode {
+                Op::CompositeConstruct => Some((inst.result_id.unwrap(), inst.clone())),
+                Op::CompositeInsert if inst.operands.len() == 3 => {
+                    Some((inst.result_id.unwrap(), inst.clone()))
+                }
+                _ => None,
+            })
+            .collect();
+        let mut unused: FxHashSet<_> = reference.keys().copied().collect();
+        for inst in function.all_inst_iter_mut() {
+            if inst.class.opcode == Op::CompositeExtract && inst.operands.len() == 2 {
+                let mut composite = inst.operands[0].unwrap_id_ref();
+                let index = inst.operands[1].unwrap_literal_int32();
+
+                let origin = loop {
+                    if let Some(inst) = reference.get(&composite) {
+                        match inst.class.opcode {
+                            Op::CompositeInsert => {
+                                let insert_index = inst.operands[2].unwrap_literal_int32();
+                                if insert_index == index {
+                                    break Some(inst.operands[0].unwrap_id_ref());
+                                }
+                                composite = inst.operands[1].unwrap_id_ref();
+                            }
+                            Op::CompositeConstruct => {
+                                break inst.operands.get(index as usize).map(|o| o.unwrap_id_ref());
+                            }
+                            _ => unreachable!(),
+                        }
+                    } else {
+                        break None;
+                    }
+                };
+
+                if let Some(origin_id) = origin {
+                    rewrite_rules.insert(
+                        inst.result_id.unwrap(),
+                        rewrite_rules.get(&origin_id).map_or(origin_id, |id| *id),
```

Comment:
Hmm. I haven't convinced myself that this is correct, but my head is mush right now and I could easily be missing something.

I believe this is to address "transitive" rewrites - something like the following code pattern:

```rust
let a = A { val: x }; // CompositeConstruct
let y = a.val; // CompositeExtract
let b = B { val: y }; // CompositeConstruct
let z = b.val; // CompositeExtract
```

In this code, the `reference` map would contain `a[0] = x` and `b[0] = y`. When processing the first CompositeExtract, `rewrite_rules` would contain the rewrite `y -> x`. When processing the second CompositeExtract, it would first seem to insert `z -> y`, but the statement this comment is on would find that `y` is already in the map, and actually insert the rewrite rule `z -> x`, as it should.

But, imagine the statements are processed in the opposite order - which could happen if it's in a loop, or some other wacky construction of the CFG. In that case, it would first insert the `z -> y` rewrite, and then insert the `y -> x` rewrite. That would mean that `y` would get rewritten to `x`, as intended, but `z` would get rewritten to the now floating-in-the-void `y` variable instead of `x`.

I think the easiest way around this would be to change this line to a plain `rewrite_rules.insert(inst.result_id.unwrap(), origin_id)` statement, and then after the information gathering phase is over, do a, uh, "transitive closure computation" (idk words), something like this (pseudocode):

```
for entry in rewrite_rules:
    while rewrite_rules.contains(entry.value):
        entry.value = rewrite_rules[entry.value]
```

that way the information gathering phase is kept simple, and there's no weird edge cases around which order rules are inserted in, but the transitive closure still happens.

Reply:
I'm pretty sure a loop would go through some OpPhis here?.. But I agree it's a sus place.
I've tried transitive closure computation - it's quite slow by itself (and also requires cloning the keys since you cannot really borrow the map both ways).
I'll try doing BOTH the eager replacement and closure computation, and profile that.

Reply:
Resolved - included the closure calculation as verbatim as one might.